### PR TITLE
Feature/disable monitor

### PR DESCRIPTION
**Pull request by DimCitus**
Not having a monitor has some impact in the internals of the code, which assumes we need to have the monitor URI and be able to contact it. This patch has some refactoring to stop assuming we should have a monitor to talk to.

### CODE REVIEW

**Comment by marcocitus** [comment]
Was wondering whether I could step through the FSM after starting with --disable-monitor, but got a segmentation fault.

```
$ pg_autoctl create postgres --pgdata node1 --pgport 6001 --disable-monitor  
$ pg_autoctl do fsm assign single --pgdata node1
$ pg_autoctl do fsm step --pgdata node1
03:53:41 INFO  Managing PostgreSQL installation at "node1"
03:53:41 INFO  Calling node_active for node /0/0 with current state: PostgreSQL is running is false, sync_state is "", latest WAL LSN is .
Segmentation fault
```

**Comment by DimCitus** [comment]
Good spot! You're not supposed to use `pg_autoctl do fsm step` when the monitor is disabled, because all the command does is ask the monitor what's the next step. So now we have:

```
$ PG_AUTOCTL_DEBUG=1 ../temp/pg_auto_failover/src/bin/pg_autoctl/pg_autoctl do fsm step --pgdata /tmp/sm/a
16:56:04 INFO  Managing PostgreSQL installation at "/tmp/sm/a"
16:56:04 FATAL The command `pg_autoctl do fsm step` is meant to step as instructed by the monitor, and the monitor is disabled.
16:56:04 INFO  HINT: see `pg_autoctl do fsm assign` instead
```

**Comment by mtuncer** [comment]
create postgres command no longer creates the db instance after this

I run to create data node

```
pg_autoctl create postgres --pgdata p2 --pgport 5000 --nodename localhost --disable-monitor
```

It prints following text and does not create data files

```
15:05:28 INFO  Found pg_ctl for PostgreSQL 11.2 at /usr/local/pgsql/bin/pg_ctl
15:05:28 INFO  Writing keeper init state file at "/Users/mtuncer/.local/share/pg_autoctl/Users/mtuncer/dev/ha/test/p2/pg_autoctl.init"
15:05:28 INFO  Keeper has been succesfully initialized.
```
it does not get created. But when I run an fsm command
```
pg_autoctl do fsm step single --pgdata p1
````
It creates the db files and does some initialization as follows

```
15:04:57 INFO  Managing PostgreSQL installation at "p1"
15:04:57 FATAL The command `pg_autoctl do fsm step` is meant to step as instructed by the monitor, and the monitor is disabled.
15:04:57 INFO  HINT: see `pg_autoctl do fsm assign` instead
mtuncer@Murat's Mac test % pg_autoctl create postgres --pgdata p2 --pgport 5000 --nodename localhost --disable-monitor
15:05:28 INFO  Found pg_ctl for PostgreSQL 11.2 at /usr/local/pgsql/bin/pg_ctl
15:05:28 INFO  Writing keeper init state file at "/Users/mtuncer/.local/share/pg_autoctl/Users/mtuncer/dev/ha/test/p2/pg_autoctl.init"
15:05:28 INFO  Keeper has been succesfully initialized.
mtuncer@Murat's Mac test % pg_autoctl do fsm assign single --pgdata p2                                                
15:06:44 INFO  Managing PostgreSQL installation at "p2"
15:06:44 INFO  FSM transition from "init" to "single": Start as a single node
15:06:44 INFO  Initialising postgres as a primary
15:06:48 INFO  Initialising a PostgreSQL cluster at "p2"
15:06:48 INFO  Now using absolute pgdata value "/Users/mtuncer/dev/ha/test/p2" in the configuration
15:06:48 INFO  Postgres is not running, starting postgres
15:06:48 INFO   /usr/local/pgsql/bin/pg_ctl --pgdata /Users/mtuncer/dev/ha/test/p2 --options "-p 5000" --options "-h *" --wait start
15:06:48 INFO  CREATE DATABASE postgres;
15:06:48 INFO  The database "postgres" already exists, skipping.
15:06:48 INFO  Transition complete: current state is now "single"
{
    "postgres": {
        "pgdata": "\/Users\/mtuncer\/dev\/ha\/test\/p2",
        "pg_ctl": "\/usr\/local\/pgsql\/bin\/pg_ctl",
        "version": "11.2",
        "host": "\/tmp",
        "port": 5000,
        "proxyport": 0,
        "pid": 39904,
        "in_recovery": false,
        "control": {
            "version": 1100,
            "catalog_version": 201809051,
            "system_identifier": "6766183265680395208"
        }
    },
    "state": {
        "current_role": "single",
        "assigned_role": "single",
        "version": 1,
        "groupId": -1,
        "nodeId": -1
    }
}
```


On top of that, I can't run pg_autoctl run on the created cluster
```
pg_autoctl run --pgdata p2
15:09:23 INFO  Managing PostgreSQL installation at "p2"
15:09:23 INFO  pg_autoctl service stopping
```


Is this expected ?

**Comment by mtuncer** [comment]
another one
I tried recently introduced get/set methods

```
pg_autoctl get node candidate-priority  --pgdata p1
15:52:53 INFO  Managing PostgreSQL installation at "p1"
15:52:53 ERROR Failed to parse connection string "PG_AUTOCTL_DISABLED": missing "=" after "PG_AUTOCTL_DISABLED" in connection info string
```

We should have a better looking error message

**Comment by DimCitus** [comment]
> I tried recently introduced get/set methods
> We should have a better looking error message

Indeed. I made the changes with the other commands, and those are more recent so I didn't cover them in the first pass. Done now, thanks!

**Comment by mtuncer** [comment]
I am getting segmentation fault on
`pg_autoctl get formation number-sync-standbys --pgdata  p1`

looks like we should also error out in `cli_get_formation_number_sync_standbys` when run in disable-monitor mode

**Comment by mtuncer** [comment]
we can create the node with --disable-monitor in regular (non-debug) mode however, setting a startup state requires being on debug mode. (PG_AUTOCTL_DEBUG=1)

**Comment by DimCitus** [comment]
> I am getting segmentation fault on
> `pg_autoctl get formation number-sync-standbys --pgdata p1`
> 
> looks like we should also error out in `cli_get_formation_number_sync_standbys` when run in disable-monitor mode

I had another look at all the functions in cli_get_set_properties.c and realised that the missing pgdata and Postgres not running conditions were not always set the same way, and also the errors reading the configuration file were not always processed the same way. I fixed that now, which also fixes the situation you are reporting.

File: src/bin/pg_autoctl/keeper_config.h [hunk 1]
```diff
@@ -50,19 +50,24 @@ typedef struct KeeperConfig
...
 void keeper_config_init(KeeperConfig *config,
-						bool missing_pgdata_is_ok,
-						bool pg_is_not_running_is_ok);
+						bool missingPgdataIsOk,
+						bool pgIsNotRunningIsOk);
 bool keeper_config_read_file(KeeperConfig *config,
-							 bool missing_pgdata_is_ok,
-							 bool pg_not_running_is_ok);
-bool keeper_config_read_file_skip_pgsetup(KeeperConfig *config);
+							 bool missingPgdataIsOk,
```
Comment (mtuncer) [inline src/bin/pg_autoctl/keeper_config.h]: wonder if we consider a bitmap flag here instead of 3 booleans.

**Pull request merged**